### ISSUE TITLE
How to handle all the ASR backends?

### ISSUE THREAD

**Issue by certik** [lfortran/lfortran#816]
We currently have the following backends:

* Lowering backends: LLVM, WebAssembly, x86
* Language backends: C, C++, Julia

For lowering backends, I also want to add WebAssembly->x86 and arm, and if the speed is good enough, perhaps just maintain two lowering backends: LLVM and WebAssembly; and do WebAssembly->x86/arm. We might also add some other engines, like [qbe](https://c9x.me/compile/). In general though, I expect just a handful of backends there, as the majority will be funneled via LLVM or WebAssembly.

For the language backends on the other hand, we need to add: Fortran, Python; then other languages that people might like: Rust, Go, JavaScript, Swift, Haskell, C#, Java, JVM, Lua, PHP, etc.

There can easily be dozens of backends.

For each backend we need thorough testing, as strings, as well as via the actual language compiler/interpreter. 

All these backends are strictly ASR->language, they are not tied (or should not be) to Fortran, or LFortran.

We need to figure out a solution that will not slowdown our fast development of LFortran, when we are trying to bring it from alpha to beta, that is, mostly adding Fortran AST->ASR semantics, as well as ASR->LLVM lowering.

One solution is to have each backend in its own repository. Since the backend work does not require modifying ASR, we can simply depend on libasr as a third party library (via a git submodule or otherwise). Since we also need LFortran to convert Fortran files to ASR for tests, we might just depend on LFortran as a third party dependency.

Then when we change ASR, we simply make LFortran working, and ensure all backends inside the LFortran tree work. Then we go to the Rust repository, update LFortran and fix all issues in the backend and ensure all tests pass.

How to ship this to end users? Say, how to create Conda packages.

One option is to create plugins. So the backend can be a plugin, that the LFortran frontend can register, and expose via command line options. Then you can do `--show-rust` (or whatever) and everything would just work; but at the same time the source code is maintained in a separate repository, thus not slowing the main LFortran development work down.

**Comment by certik** [comment]
If we go the plugin route, we want the plugin to work both with LFortran and LPython, as well as any other future frontend that we write to LCompilers.

**Comment by lucifer1004** [comment]
I think the plugin route is the right choice in the long run, but a solid framework is needed to make that work and it will take much effort.

**Comment by lucifer1004** [comment]
Also, we need to rethink the organization's structure. Seems that all front-ends and back-ends should be put under the LCompilers organization.

**Comment by certik** [comment]
Yes, historically we created lfortran first, only then expanded into LCompilers. Indeed, I think the frontends should also be plugins, so that you can ideally choose what frontends and what backends you need and can easily compile them in.

**Comment by lucifer1004** [comment]
Agree. The ASR is the soul, while the frontends and backends are flesh.

**Comment by czgdp1807** [comment]
All of what you said makes sense to me. 

> One solution is to have each backend in its own repository. Since the backend work does not require modifying ASR, we can simply depend on libasr as a third party library (via a git submodule or otherwise). Since we also need LFortran to convert Fortran files to ASR for tests, we might just depend on LFortran as a third party dependency.

May or may not be. Like we had to update GoTo and GoToTarget for C backend. It may happen rarely.

Also note that we need more people to add and maintain backends. Like one person should manage and maintain at most 2 backends at a time. That way no one loose focus and efficiency.

The plus point of having more backends we get connected to more communities and get more exposure, users. Also our ASR design will also get updated and get better to satisfy constraints of all the backends.

**Comment by czgdp1807** [comment]
> Since the backend work does not require modifying ASR, we can simply depend on libasr as a third party library (via a git submodule or otherwise).

There is one point to be noted. `libasr` is updating really fast, like with each PR we update the codegen, and ASR. So the problem would be change in ASR due to one backend would require all the other backends to be updated. It may be rare case but still. Also, modifying or deleting any node of ASR will require all the backends to be updated. 

These issues won't be there once `libasr` stabilises.

**Comment by lucifer1004** [comment]
> `libasr` is updating really fast.

I agree. So I suggest not splitting the repo for now.

**Comment by certik** [comment]
Let's collect the facts:

* The following backends are all under serious consideration: Fortran, Python, Rust, Go, JavaScript, Swift, Haskell, C#, Java, JVM, Lua, PHP, Scheme, and more.

* ASR is guaranteed to change, even quite drastically; We will definitely remove and add nodes, but we will also almost for sure make big improvements to the type system etc.

* We can't put all backends into the same mono repo, that would ensure we fail to deliver on LFortran. Even a tiny change to ASR would take several hours to update all the backends, just to get the change into LFortran (or LPython). We don't have the team currently to do that.

The only way forward is to have each backend in its own repository, just like each frontend (currently LFortran, LPython) is in its own repository.

What remains to solve is how to distribute this to make it easy to use for end users. We make it easy to build LFortran with any of the backends in separate repositories. Then when we do a release, all we need is to update the Rust backend (let's say) to work with the release and document which version to use. The Conda LFortran package can build all the backends.

So we need to just make the Rust backend repository an optional dependency in LFortan's cmake build system. We can add the command line option like `--show-rust`, and it would say you need to rebuild with Rust support, just like the `--show-llvm` option does if you build without LLVM.

**Comment by lucifer1004** [comment]
@certik I think we may learn from the design of `pandoc`: https://pandoc.org/filters.html

Under the hood it is very similar to what we are doing here:

- pandoc: INPUT --reader--> AST --filter--> AST --writer--> OUTPUT
- LCompilers: INPUT --frontend--> AST --> ASR --pass--> ASR --backend--> OUTPUT

**Comment by certik** [comment]
The Pandoc is a good example. The readers are 1500 - 2000 lines: https://github.com/jgm/pandoc/blob/ad0532244adfbd0ca9596a27506ed6161e4bc225/src/Text/Pandoc/Readers/RST.hs, the writers seem to be a lot shorter, around 800 lines: https://github.com/jgm/pandoc/blob/ad0532244adfbd0ca9596a27506ed6161e4bc225/src/Text/Pandoc/Writers/Markdown.hs.

In comparison, LPython parser is 1250 lines just the Bison grammar, 700 lines tokenizer (although with a large comment), 1000 lines converting from Bison to AST. The AST to ASR is currently about 7000 lines. I skipped some smaller utility files that I think will later be refactored into libasr. LPython is still in alpha, I expect it to at least double when it reaches production. 

The LFortran's parser is at about 9000 lines (not counting any generated files), this includes both free-form (pretty much complete) and fixed-form (almost can parse SciPy). The AST to ASR is about 8000 lines currently, and when all is done, I am guessing it might double.

So overall the frontend is about 10x larger than Pandoc's frontends.

Regarding backends, LFortran's AST -> SRC is 4400 lines and should be complete. The ASR->LLVM is about 8000 lines and it will probably still double. The ASR->C is currently 2000 lines, but it will probably also at least double.

So backends are also at least 10x bigger than Pandoc's backends.

Regarding ASR->AST->backend, I thought about this, and initially I used to do that, but lately we just generate source code directly. For Fortran, ASR->AST would be at least 2000 lines, probably more, so ASR->AST->SRC would be over 6000 lines. A direct ASR->SRC backend might be only around 2000 or so lines, probably not more than the AST->SRC. So if we just want to generate the target language, it's better to do ASR->SRC directly, less layers to maintain, most people understand the source code, etc.  If we had to maintain AST->SRC (such as we already have it in LFortran), then it's not clear, but even then it seems it's probably easier to do ASR->SRC. However, one idea that I just got is that if we use a different language to do the ASR->AST transformation (which is just list rewriting), such as in Clojure, and somehow we generated the C++ visitor from it, it might be quite easy.

Finally, there is the standard library, which is generally huge (especially for Python). Pandoc doesn't have it.

Overall, we are dealing with at least an order of magnitude more code for everything, so I would keep things separate in separate repositories. Definitely the frontends, but I think also backends.

I would say, let's start with separate Rust backend, and we can later decide if it would make sense to merge it to libasr, I think that would be relatively straightforward.

**Comment by lucifer1004** [comment]
By mentioning `pandoc`, I did not mean that we should put all frontends and backends together. Instead, I meant that we could learn something from its design.

It is true that `pandoc` embeds many readers and writers, but it is also possible to write custom readers and writers:

- https://pandoc.org/custom-writers.html
- https://pandoc.org/custom-readers.html

I think writing new frontends and backends for LCompilers can be as easy as this. And that would be charming.

**Comment by certik** [comment]
Especially in more higher level language like Python. C++ is a bit more verbose and heavy duty, but we need it for performance of compilation.

**Comment by lucifer1004** [comment]
Yes. For example, `pandoc` itself is written in Haskell, but custom readers/writers can be written in Lua.

**Comment by lucifer1004** [comment]
And just like `pandoc`'s CLI command

`pandoc -f readable.lua -t markdown https://pandoc.org`

ideally, we can have a similar CLI command

`lcompiler -f python -t julia test.py`

**Comment by meow464** [comment]
1. `share/lfortran/backends/config/name` is a text file and maps the backend library filename to metadata (plugin system version, backend name, architecture, compiler used, lfortran version, function names i.e. `asr_to_rust` and `emit_rust`, etc).
2. `lfortran --list-backends` scans `share/lfortran/backends/config/*` and lists all backends compatible with this lfortran version.
3. `lfortran --backend rust` uses `dlopen` and `dlsym` to open the dynamic library and load `asr_to_rust()` and `emit_rust()`

Note: `dlsym()` takes a symbol name and thus `asr_to_rust()` and `emit_rust()` would have to be declared as `extern "C"` to disable name mangling but would otherwise work the same. 

Building: One would just checkout the backend submodule with cmake and compile with `cmake --build . --target rust_target`.
Distributing: Simply distribute through conda or any other package manager and install the necessary files under `share/lfortran/backends/`

Problems: libasr is not a real library ATM and thus backends would include duplicated libasr code. Hopefully this won't be a problem if backends target a specific lfortran commit.

Code changes: Some changes to lfortran.cpp and auxiliary functions for scanning and loading plugins. Should take less effort than writing the rust backend.
